[Tutorial 5] MVVMを学ぶ

k-shimizu3 opened the pull request:
### 課題
 - Tutorial 5
 - MVVMアーキテクチャに触れて、手元に動くソースを作る
 - 各パッケージの導入
   - freezed
   - json_serializable
   - riverpod
   - retrofit

---

### 主な実装内容
Tutorial 5のNotionに記載されている参考リポジトリと同じ内容です。
（パブリックリポジトリのため、NotionやリポジトリのURLの記載は省略させていただきます🙇🏻‍♂️）

-  lib/main.dart
   - ProviderScopeの追加
   - qiita_client_screen.dartへの遷移ボタンを追加

 - lib/qiita/ 
    - api/
    - model/
    - repository/
    - state/
    - qiita_client_screen.dart
    - qiita_client_state_notifier.dart

 - pubspec.yaml
    - パッケージの追加

---
### 動作サンプル
https://user-images.githubusercontent.com/114896398/196843161-dcbb716c-8d0d-4be8-bf19-9cc7959dfe0a.mov

---

### 課題（悩んだこと）
特にありません。

---

### 備考
いつもレビューありがとうございます。
レビューよろしくお願いします！

k-shimizu3 committed:
QiitaClientScreenへの遷移ボタンの追加

k-shimizu3 committed:
パッケージの追加

k-shimizu3 committed:
お手本を参考に実装

k-shimizu3 committed:
自動生成ファイルの追加

onishitaisei commented on lib/qiita/model/qiita_item.dart:
```diff
@@ -0,0 +1,18 @@
+@freezed
+class QiitaItem with _$QiitaItem {
+  const factory QiitaItem({
+    @JsonKey(name: "title") String? title,
```
`@JsonKey`はプロパティ名と返却されたkey名が違う場合に使うものだと思うので、今回の場合はいらないかなと思いました！

よく使う例だと、プロパティがキャメルケース（userId）で返却されたkey名がスネークケース（user_id）、これを合致させるために使う場合が多いかなと思います。

k-shimizu3 replied:
修正しました！
`@JsonKey`を何となく使ってしまっていたのでとても勉強になりました！

onishitaisei commented on lib/qiita/model/qiita_user.dart:
```diff
@@ -0,0 +1,14 @@
+@freezed
+class QiitaUser with _$QiitaUser {
+  const factory QiitaUser({
+    @JsonKey(name: "profile_image_url") String? profileImageUrl,
```
上記でコメントした例はまさにこの部分です！

k-shimizu3 replied:
ありがとうございます！

onishitaisei commented on lib/qiita/qiita_client_screen.dart:
```diff
@@ -0,0 +1,103 @@
+    return
+        // 戻るボタンの挙動
+        WillPopScope(
+      onWillPop: state.isReadyData
```
個人的にこのコメントだったら`onWillPop`プロパティ上に書いた方がわかりやすいかなと思いました！

k-shimizu3 replied:
修正しました！

onishitaisei commented on lib/qiita/qiita_client_screen.dart:
```diff
@@ -0,0 +1,103 @@
+            ),
+            Visibility(
+              visible: state.isLoading,
+              child: Container(
```
colorプロパティを使用するだけであればContainerではなく`ColoredBox`を使った方が目的がはっきりしてるので良いかもです！

k-shimizu3 replied:
修正しました！
`ColoredBox`のことを知らなかったので勉強になります！📝

onishitaisei commented on lib/qiita/qiita_client_screen.dart:
```diff
@@ -0,0 +1,103 @@
+        final item = qiitaItems[index];
+
+        return Container(
+          padding: EdgeInsets.only(top: 4, bottom: 4, left: 8, right: 8),
```
上下、左右それぞれが同様の値を設定するのであれば`EdgeInsets.symmetry(vertical: 4, horizontal:8, );`の方が簡潔に書けるので良いかもです！

k-shimizu3 replied:
修正しました！

onishitaisei commented on lib/qiita/qiita_client_state_notifier.dart:
```diff
@@ -0,0 +1,43 @@
+    final qiitaItems =
+        await _read(qiitaRepositoryProvider).fetchQiitaItems(tag);
+
+    if (qiitaItems.length != 0) {
```
配列の長さチェックは`isEmpty`プロパティを使用した方が直感的なので良いかもです！

k-shimizu3 replied:
ありがとうございます！
修正しました！

onishitaisei commented:
@kenji-axia 
数件コメントしました！全体的に良く書けていると思います🙆‍♂️

kenji-axia commented:
@onishitaisei
レビューありがとうございます！🙇🏻‍♂️
コメント頂いた箇所を修正（https://github.com/kenji-axia/flutter_tutorial/pull/10/commits/08f2d84db227ee0cfb70158cda0b99d6960f5a95 ）しましたので、お手隙の際にご確認お願いします。

k-shimizu3 committed:
レビュー修正

修正点：
不要なJsonKeyの削除
コメント位置を修正
colorプロパティしか使っていないContainerをColoredBoxに変更
EdgeInsetsの修正
配列の空チェックを" .isNotEmpty "に変更

onishitaisei commented:
@kenji-axia 
LGTM!!

k-shimizu3 committed:
developブランチをマージ

先ほどGitHub上で feature/tutorial4ブランチを developブランチへマージしたので、その分をこのブランチにも反映させました。

* develop:
  _AsyncScreenStateのメソッド名の変更
  コードの整理
  Tutorial4の実装
  shared_preferences: ^2.0.12 の追加
  データクラスの追加

# Conflicts:
#	lib/main.dart
#	pubspec.yaml

t-nakata reviewed (approved):
@onishitaisei 
approve忘れてますよ

@kenji-axia 
LGTMなんだけど

これをマージすると後続のPRにコンフリクトが発生するかもしれないので、そのときは注意してね！

kenji-axia commented:
@t-nakata 
ありがとうございます！
注意してマージしようと思います！
